### PR TITLE
Hotend Protection - Prevent deep-fried filament clogs

### PR DESCRIPTION
**Pull request by italocjs**
### Description

This feature has been developed to prevent clogs and jams caused by filament carbonization, usually caused by a nozzle being heated for too much time without extrusion (usually as result of someone preheating and forgetting)

Logic:
It checks if extruder have not moved during FCP_TIMEOUT(in minutes) and check if temperature is above FCP_MIN_TRIGGER(celsius), if both conditions are met, then it will set the nozzle and bed to their respective FCP_NOZZLE_TARGET and FCP_BED_TARGET
 
New files:
\features\FCP.h
\features\FCP.cpp

Changed:
\marlin.cpp (added the "call" for the funcion)
\configurations.h (added the definitions line from 2216 to 2230) and used my printer config to test. 

### Benefits

Prevent user-caused clog and jams

### Future improvements
1 - As said in the "best pratices" i did not managed to make use of the elapsed function (to compare time) duo overflows, so i wrote the time comparison code another way.  

### Related Issues
#16200

### CODE REVIEW

**Comment by extesy** [comment]
@italocjs You should separate changes that are necessary for this feature from the rest of the changes that are specific for your printer.

**Comment by thinkyhead** [comment]
Thanks for your contribution! Unfortunately we can't accept PRs directed at release branches. We make patches to the bugfix branches and only later do we push them out as releases. Please redo this PR starting with the `bugfix-2.0.x` branch and be careful to target `bugfix-2.0.x` when resubmitting the PR.

It may help to set your fork's default branch to `bugfix-2.0.x`.

See http://marlinfw.org/docs/development/getting_started_pull_requests.html for full instructions.

File: Marlin/Configuration.h [hunk 1]
```diff
@@ -2215,3 +2213,18 @@
...
+ * 
+ * Its not a good idea to keep the timeout less than 15min, as this is a conservative bed heating time
+*/
+#define FCP_ENABLE
+#if ENABLED(FCP_ENABLE)
+  #define FCP_TIMEOUT 0.2 // MAX Time without extruder movement to trigger the protection (min)
+  #define FCP_MIN_TRIGGER 180 //MIN Temp to trigger FCP
+  #define FCP_NOZZLE_TARGET 0 // (ºC) Target for the nozzle E0 and E1
+  #define FCP_BED_TARGET 0 //  (ºC) Target for the bed 
+#endif
```
Comment (thinkyhead) [inline Marlin/Configuration.h]: This added block is good, but all other changes must be removed from this file before consideration.

**Comment by thinkyhead** [comment]
The basic concept is sound, but the heuristic might be too simple and the behavior could be problematic when combined with other behaviors. For example, if the user does an intentional pause from the host, the temperature will not be returned to the previous state before continuing the job. Ideally, this should invoke a procedure like the current "pause" (or most of it) in which the screen is locked while the machine is in a temperature standby, and there is a prompt to re-heat and continue.

**Comment by italocjs** [comment]
Hi @thinkyhead , sorry for the delay.  Thanks for the fixes in the code, sorry i submitted to the wrong branch, i tried following that guide but i'm not sure where i did wrong.

Anyway,  i agree with the pause situation, i'll try studying how the pause works and implement something to fix this.

**Comment by italocjs** [comment]
Sorry, i'm quite new to github/programming, but was this merged into the bugfix? i cant find the feature there, if not, i noticed that the last message says something about conflicts, its because my config.h has my own configs?

**Comment by italocjs** [comment]
Another thing i forgot to mention, i have been printing since i developed this (and thinkyhead improved) without any issue,  it has worked well and only detected when i truly forgot

**Comment by thinkyhead** [comment]
Have you tried the existing `EXTRUDER_RUNOUT_PREVENT` feature? Maybe the changes here should be merged with that feature instead of making it separate.

**Comment by bill-orange** [comment]
Is the incorporation of the feature on the road map?  It would provide an backstop to my forgetfulness.

**Comment by thisiskeithb** [comment]
Renaming this to “Heater Timeout” or similar (like the [OctoPrint plugin](https://plugins.octoprint.org/plugins/heatertimeout/)) might be a better name.

It’d be a nice safety feature.

**Pull request merged**